Check resource generation when processing updates of some resources to skip config regeneration

kevin85421 opened the pull request:
### Proposed changes

Because I lack sufficient context for this repository, I decided to approach this PR step-by-step and ensure each step is correct.

## Requirement 1: Move generation check from ChangeProcessor to Controllers, for the types we already check generation

* **Step 1: Move generation check from ChangeProcessor to Controllers, for the types we already check generation** https://github.com/nginxinc/nginx-gateway-fabric/pull/1422/commits/0d20c6185412f3cf2ae94f4c5450f0885d4b148f
  * Remove `type generationChangedPredicate struct{}`. (Update: Add it back in Step 2. See Step 2 for more details.)
  * In `change_processor.go`, there are four kinds of Kubernetes resources that use `generationChangedPredicate`. These include **GatewayClass**, **Gateway**, **HTTPRoute**, and **ReferenceGrant**.
  * In this commit, I only add `controller.WithK8sPredicate(k8spredicate.GenerationChangedPredicate{})` for these four Kubernetes resources in `manager.go` in this commit.

* **Step 2: Remove tests that do not involve a `generation` change from `change_processor_test.go`.** (https://github.com/nginxinc/nginx-gateway-fabric/pull/1422/commits/9e01812d03be3e8cd0d7b77b493486fa7e397159)
  * `change_processor.go` should not handle any cases with the same generation, as these should already be filtered out by the controller's `GenerationChangedPredicate`.
  * In Step 1, I removed `type generationChangedPredicate struct{}` and replaced the `predicate` for the four Kubernetes resources in `change_processor.go` with `nil`. However, `predicate: nil` indicates that the state change will not trigger a graph rebuild. See #1358 for more details. This behavior is incompatible with the old behavior. Hence, I added `type generationChangedPredicate struct{}` back and made it always return `true` to keep the behavior compatible. We may need to remove it in the future.

## Requirement 2: Figure out which types miss generation check and add it.

TL;DR: There is no Kubernetes resources that miss generation check.

"Requirement 1" is entirely equivalent to the existing behavior, and "Requirement 2" introduces new behavior. 

* `GatewayClass`: The existing `GatewayClassPredicate` filters out all events that are not using NGF as the controller by checking `Spec.ControllerName`.  Add `k8spredicate.GenerationChangedPredicate` to `GatewayClass`. See https://github.com/nginxinc/nginx-gateway-fabric/pull/1422#discussion_r1440957893 for more details.
* `Gateway`: Add `k8spredicate.GenerationChangedPredicate` to `Gateway`. See https://github.com/nginxinc/nginx-gateway-fabric/pull/1422#discussion_r1440957893 for more details.
* `HTTPRoute`: Add `k8spredicate.GenerationChangedPredicate` to `HTTPRoute`. The behavior remains equivalent, with or without this PR. See "Requirement 1" for more details.
* `Service`:
  * The existing `ServicePortsChangedPredicate` filters out all events that do not have changes in `Spec.Ports`. If `ServicePortsChangedPredicate` is true, `k8spredicate.GenerationChangedPredicate` should always be true. Hence, we don't need to add `k8spredicate.GenerationChangedPredicate`.
  * The existing `GatewayServicePredicate` filters out all events without changes in `Spec.Type` and `Status.LoadBalancer.Ingress`. As I understand it, changes in `Status` do not increment the generation value. Therefore, `k8spredicate.GenerationChangedPredicate` might filter out some changes related to `Status.LoadBalancer.Ingress`. Hence, we should not add the predicate for this case.
  * It seems that the Kubernetes service does not have a `metadata.generation` field. I am not 100% sure, but it's highly likely.
    ```sh
    kubectl get svc $SVC_NAME -o yaml | grep generation # nothing should be shown.
    ```
* `Secret`: 
  * It seems that the Kubernetes Secret does not have a `metadata.generation` field. I am not 100% sure, but it's highly likely.
    ```sh
    kubectl get secret $SECRET_NAME -o yaml | grep generation # nothing should be shown.
    ```
* `EndpointSlice`: The `k8spredicate.GenerationChangedPredicate` already exists.
* `Namespace`: 
  * It seems that the Kubernetes Namespace does not have a `metadata.generation` field. I am not 100% sure, but it's highly likely.
    ```sh
    kubectl get namespace $NS_NAME -o yaml | grep generation # nothing should be shown.
    ```
* `CRD`
  * The existing `AnnotationPredicate` filters out all events without changes in `metadata.Annotations`. Changes to annotations on a custom resource will not increment the `metadata.generation`. Therefore, `k8spredicate.GenerationChangedPredicate` might filter out events that only involve changes in `metadata.Annotations`.

Closes #825 

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [ ] I have read the [CONTRIBUTING](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that all unit tests pass after adding my changes
- [ ] I have updated necessary documentation
- [ ] I have rebased my branch onto main
- [ ] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

kevin85421 commented on internal/mode/static/manager.go:
```diff
@@ -265,22 +265,28 @@ func registerControllers(
 		{
 			objectType: &gatewayv1.Gateway{},
 			options: func() []controller.Option {
+				options := []controller.Option{
```
I want to double-check with the maintainers: Is it correct to add the `GenerationChangedPredicate` option regardless of the value of `GatewayNsName`?

sjberman replied:
I think that's right. We still see other Gateways and keep track of them.

sjberman replied:
If our current behavior is to only watch the Gateway we care about, I think we would only want predicates to apply in that case.

kate-osborn replied:
The predicates and `NamespacedNameFilter` are applied in different places. Predicates are applied before the controller's `Reconcile` method is called, and the `NamespacedNameFilter` is applied in the `Reconcile` method. So, we can't selectively apply the predicate to a single Gateway without refactoring the code and changing the `NamespacedNameFilter` to a predicate. 

I think what you have here works.  Basically, we first filter out all Gateway upsert events where the generation hasn't changed. Then, in `Reconcile` we apply the `NamespacedNameFilter` and further filter out the Gateways that don't match the `GatewayNsName`.

sjberman replied:
Thanks for the clarification @kate-osborn

kevin85421 commented on internal/mode/static/state/change_processor_test.go:
```diff
@@ -1737,14 +1695,7 @@ var _ = Describe("ChangeProcessor", func() {
 			})
 
-			It("should report not changed after multiple Upserts of unrelated and unchanged resources", func() {
-				// unchanged Gateway API resources
```
https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/predicate#GenerationChangedPredicate

Because the resources are unchanged, the `generation` should remain the same and be filtered out by the controller's `GenerationChangedPredicate`.

kevin85421 commented on internal/mode/static/state/change_processor_test.go:
```diff
@@ -703,17 +703,6 @@ var _ = Describe("ChangeProcessor", func() {
 					Expect(helpers.Diff(expGraph, graphCfg)).To(BeEmpty())
 				})
 			})
-			When("the first HTTPRoute without a generation changed is processed", func() {
```
`change_processor.go` should not handle any cases with the same generation, as these should already be filtered out by the controller's `GenerationChangedPredicate`.

kevin85421 commented on internal/mode/static/state/change_processor_test.go:
```diff
@@ -733,17 +722,6 @@ var _ = Describe("ChangeProcessor", func() {
 				},
 				)
 			})
-			When("the first Gateway update without generation changed is processed", func() {
```
`change_processor.go` should not handle any cases with the same generation, as these should already be filtered out by the controller's `GenerationChangedPredicate`.

kevin85421 commented on internal/mode/static/state/change_processor_test.go:
```diff
@@ -758,17 +736,6 @@ var _ = Describe("ChangeProcessor", func() {
 					Expect(helpers.Diff(expGraph, graphCfg)).To(BeEmpty())
 				})
 			})
-			When("the GatewayClass update without generation change is processed", func() {
```
`change_processor.go` should not handle any cases with the same generation, as these should already be filtered out by the controller's `GenerationChangedPredicate`.

kevin85421 commented on internal/mode/static/state/change_processor_test.go:
```diff
@@ -1590,15 +1557,6 @@ var _ = Describe("ChangeProcessor", func() {
 				changed, _ := processor.Process()
 				Expect(changed).To(BeTrue())
 			})
-			It("should report not changed after multiple Upserts of the resource with same generation", func() {
```
`change_processor.go` should not handle any cases with the same generation, as these should already be filtered out by the controller's `GenerationChangedPredicate`.

bjee19 commented:
Hey @kevin85421, saw your comment [here](https://github.com/nginxinc/nginx-gateway-fabric/issues/825) and just wanted to give you a heads up that the majority of the engineers on the team are currently on holiday until January 2nd, @sjberman may be able to take a look tomorrow, but if not, this may just sit for a little while. Hope that's alright!

kevin85421 commented:
@bjee19 Thanks for letting me know! Happy holidays!

sjberman commented:
I'd just use on PR for all of these changes, and ensure the title of the PR defines what the high level fix is.

sjberman commented:
At first glance I think the approach looks good.

kate-osborn commented on internal/mode/static/manager.go:
```diff
@@ -265,22 +265,28 @@ func registerControllers(
 		{
 			objectType: &gatewayv1.GatewayClass{},
 			options: []controller.Option{
+				controller.WithK8sPredicate(k8spredicate.GenerationChangedPredicate{}),
```
The controller options logic will also have to be updated to support setting multiple predicates: https://github.com/nginxinc/nginx-gateway-fabric/blob/7639fe5bc8e8d4af3cc0778887042e1c887cfa90/internal/framework/controller/register.go#L44

Alternatively, you could use the [And](https://github.com/kubernetes-sigs/controller-runtime/blob/7f316f1309b11c24aa8a96750e694f897555d348/pkg/predicate/predicate.go#L236C1-L236C1) controller-runtime function to create a composite predicate.

kevin85421 replied:
Updated https://github.com/nginxinc/nginx-gateway-fabric/pull/1422/commits/cc429b2a5b6342052cc2ed2971b3230cbf17df77. I have chosen to use the And operator to create a composite predicate, to minimize the code changes required for this PR.

kate-osborn commented on internal/mode/static/state/changed_predicate.go:
```diff
@@ -31,17 +31,7 @@ type generationChangedPredicate struct{}
-
-	return newObject.GetGeneration() != oldObject.GetGeneration()
-}
+func (generationChangedPredicate) upsert(_, _ client.Object) bool { return true }
```
I recommend changing the name of this predicate to reflect its behavior. Maybe something like `alwaysTruePredicate`. 

Alternatively, if this is merged after #1432, you can remove this predicate and change the logic of the `changeTrackingUpdater` to return `true` if there's no `stateChangedPredicate` registered for the resource.

kevin85421 replied:
Updated https://github.com/nginxinc/nginx-gateway-fabric/pull/1422/commits/52063131b95c1b067b79021892ec8e035fc3cc54

kate-osborn reviewed (commented):
@kevin85421 thanks for working on this!

I left a few comments, but I think you are on the right track 👍

kevin85421 commented:
Hi @sjberman @kate-osborn, I have already addressed the comments. In addition, there are no Kubernetes resources that miss the generation check, so there is no commit for "Requirement 2". See the PR description for more details. Could you provide a pointer on where to add tests, and are there any existing tests that I can refer to? Thanks!

sjberman commented:
@kevin85421 If you could update the initial commit message to include the problem/solution structure, this will make it easier on our team when we merge it in. This way you can supply the commit message and we don't need to write one for you.

kevin85421 commented:
> @kevin85421 If you could update the initial commit message to include the problem/solution structure, this will make it easier on our team when we merge it in. This way you can supply the commit message and we don't need to write one for you.

Thanks! I have updated the commit message. I originally planned to update it after adding enough tests 😅.

kate-osborn commented on internal/mode/static/manager.go:
```diff
@@ -265,22 +265,29 @@ func registerControllers(
-					return []controller.Option{
-						controller.WithNamespacedNameFilter(filter.CreateSingleResourceFilter(*cfg.GatewayNsName)),
-					}
+					options = append(options,
```
nit: we prefer to split long lines up like this:
```
options = append(
    options,
    controller.WithNamespacedNameFilter(filter.CreateSingleResourceFilter(*cfg.GatewayNsName)),
)
    
```

kevin85421 replied:
Updated https://github.com/nginxinc/nginx-gateway-fabric/pull/1422/commits/de143ef24458dfeccd4a6c62746ac407f406ac78

kate-osborn commented on internal/mode/static/state/changed_predicate.go:
```diff
@@ -24,24 +24,12 @@ func (f funcPredicate) delete(object client.Object) bool {
-type generationChangedPredicate struct{}
-
-func (generationChangedPredicate) delete(_ client.Object) bool { return true }
+// FIXME(kevin85421): We should remove this predicate and update changeTrackingUpdater once #1432 is merged.
```
Thanks for adding a FIXME!

kate-osborn reviewed (changes requested):
@kevin85421 thanks for the details in the PR description.  

The changes look good to me, I just left one small nitpicky comment. 

As far as testing goes, we don't have any unit tests for the `registerControllers` function because they don't add a lot of value. We would just be checking that we adding the correct options for each objectType. 

In order to test this PR, I think we need to do the following: 
1. Run the conformance tests. I will kick off the pipeline which will run the conformance tests, so you don't need to run them locally.
2. Manually verify that we don't reconcile events for GatewayClasses, Gateways, HTTPRoutes, and RefGrants when the generation doesn't change. This can be verified through logs ( You'll want to look for the log that starts with "Reconciling the resource" ) I would also double check that the GatewayClass `And` predicate works as expected.


@pleshakov does this seem like a good approach or do you think it's worth it to add a unit test for the `registerControllers` function?

pleshakov commented:
@kate-osborn 
> In order to test this PR, I think we need to do the following:

> Run the conformance tests. I will kick off the pipeline which will run the conformance tests, so you don't need to run them locally.
Manually verify that we don't reconcile events for GatewayClasses, Gateways, HTTPRoutes, and RefGrants when the generation doesn't change. This can be verified through logs ( You'll want to look for the log that starts with "Reconciling the resource" ) I would also double check that the GatewayClass And predicate works as expected.
@pleshakov does this seem like a good approach or do you think it's worth it to add a unit test for the registerControllers function?

that makes sense to me

If the controllers still reconcile when there is no gen change (because of the bug), will it trigger rebuild of the graph and reload of NGINX configuration?
If that is the case,  would it also make sense to create an issue to add test automation for that in the future ? Note we don't yet have functional tests yet written but we're planning to add them in the future. I think having such tests will help along with the tests for other reload-avoiding-related optimizations we're planning to implement -- https://github.com/nginxinc/nginx-gateway-fabric/issues/1112 , https://github.com/nginxinc/nginx-gateway-fabric/issues/1123,  https://github.com/nginxinc/nginx-gateway-fabric/issues/1124

kevin85421 commented:
Thanks, @kate-osborn and @pleshakov! I have just written down how I test this PR manually. Please let me know if this is not sufficient. Thanks!

* Step 1: Install NGF, and a `GatewayClass`, which is a non-namespaced resource, will also be created.
* Step 2: Create 3 * HTTPRoute, 1 * ReferenceGrant, and 1 * Gateway by running [cafe-routes.yaml](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/examples/https-termination/cafe-routes.yaml), [gateway.yaml](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/examples/https-termination/gateway.yaml), and [reference-grant.yaml](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/examples/https-termination/reference-grant.yaml).
* Step 3: Run [this script](https://gist.github.com/kevin85421/2037a5076d00662ce55a249abbefba92) to list the information of `generation` for all GatewayClass/Gateway/HTTPRoute/ReferenceGrant.
  ```
  [Non-namespaced resource]
  -----------------------------------
  GatewayClasses:
  nginx   1
  
  
  
  [Namespaced resources]
  -----------------------------------
  Namespace: certificate
  ReferenceGrants:
  access-to-cafe-secret   1
  -----------------------------------
  Namespace: default
  Gateways:
  gateway   1
  HttpRoutes:
  cafe-tls-redirect   1
  coffee              1
  tea                 1
  -----------------------------------
  Namespace: kube-node-lease
  -----------------------------------
  Namespace: kube-public
  ```
* Step 4: Wait for a while. In my experiment, I wait for 40 minutes. Check [here](https://gist.github.com/kevin85421/0b28c8e745c3b9df773e177c2d0b600e) if you are interested in my log file.
  ```sh
  kubectl logs -n nginx-gateway my-release-nginx-gateway-fabric-67dfb666b9-jb7tk | tee log
  ```
* Step 5: As shown in Step 3, the `generation` of every related resource is 1. The number of reconciliations for each related Kubernetes resource type is equal to the count of resources of that type. Check the log file:
  * `"Reconciling the resource","controller":"httproute"` => 3
  * `"Reconciling the resource","controller":"referencegrant"` => 1
  * `"Reconciling the resource","controller":"gateway"` => 1
  * The result is as my expectation.

pleshakov commented:
Hi @kevin85421 

> Step 4: Wait for a while. In my experiment, I wait for 40 minutes. Check [here](https://gist.github.com/kevin85421/0b28c8e745c3b9df773e177c2d0b600e) if you are interested in my log file.

I wonder why we wait for 40 mins. Can we wait for less time?

I think we can avoid waiting at all if we try to change GatewayClass/Gateway/HTTPRoute/ReferenceGrant in the API without changing their generation. We can do that by adding or updating an annotation on the resource - such operation doesn't change resource generation.

For example, if I add an annotation to an HTTPRoute resource:
```
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  annotations:
    hello: hello
 ...
```

However, note that `metadata.resourceVersion` will change those after such an update.

Currently, I see the following in the logs:
```
{"level":"info","ts":"2024-01-12T17:05:31Z","msg":"Reconciling the resource","controller":"httproute","controllerGroup":"gateway.networking.k8s.io","controllerKind":"HTTPRoute","HTTPRoute":{"name":"coffee","namespace":"default"},"namespace":"default","name":"coffee","reconcileID":"35e96c96-5dff-42c6-b05c-90fdec385e70"}
{"level":"info","ts":"2024-01-12T17:05:31Z","msg":"Upserted the resource","controller":"httproute","controllerGroup":"gateway.networking.k8s.io","controllerKind":"HTTPRoute","HTTPRoute":{"name":"coffee","namespace":"default"},"namespace":"default","name":"coffee","reconcileID":"35e96c96-5dff-42c6-b05c-90fdec385e70"}
{"level":"info","ts":"2024-01-12T17:05:31Z","logger":"eventLoop","msg":"added an event to the next batch","type":"*events.UpsertEvent","total":1}
{"level":"info","ts":"2024-01-12T17:05:31Z","logger":"eventLoop.eventHandler","msg":"Handling events from the batch","batchID":27,"total":1}
{"level":"info","ts":"2024-01-12T17:05:31Z","logger":"eventLoop.eventHandler","msg":"Handling events didn't result into NGINX configuration changes","batchID":27}
{"level":"info","ts":"2024-01-12T17:05:31Z","logger":"eventLoop.eventHandler","msg":"Finished handling the batch","batchID":27}
```

Which should not happen as a result of your PR.

We can also look at Prometheus metrics  (https://docs.nginx.com/nginx-gateway-fabric/how-to/monitoring/monitoring/#controller-runtime-metrics ). Their include controller-runtime metrics. So that we can check resource reconciliation count. For example:
```
controller_runtime_reconcile_total{controller="httproute",result="error"} 0
controller_runtime_reconcile_total{controller="httproute",result="requeue"} 0
controller_runtime_reconcile_total{controller="httproute",result="requeue_after"} 0
controller_runtime_reconcile_total{controller="httproute",result="success"} 9
```
I expect the following to not change if our controllers successfully filter out generation changes.:
```
controller_runtime_reconcile_total{controller="httproute",result="success"} 9
```

kate-osborn commented:
> If the controllers still reconcile when there is no gen change (because of the bug), will it trigger rebuild of the graph and reload of NGINX configuration? If that is the case, would it also make sense to create an issue to add test automation for that in the future ?

Yes, it would trigger a rebuild and reload. I think an automated test is a good idea. I can write up an issue for it.

kevin85421 commented:
@pleshakov Thank you for the reply!

> I wonder why we wait for 40 mins. Can we wait for less time?

No. I just want to ensure the NGF controller doesn't requeue k8s resources unconditionally.

I have already updated https://github.com/nginxinc/nginx-gateway-fabric/pull/1422#issuecomment-1888612433. I added annotations to all related Kubernetes resources and checked the Prometheus metrics.

kevin85421 commented:
Gentle ping - Do the tests in https://github.com/nginxinc/nginx-gateway-fabric/pull/1422#issuecomment-1888612433 make sense to you? cc @kate-osborn @pleshakov Thanks!

kate-osborn reviewed (approved):
🚀

kate-osborn commented:
> Gentle ping - Do the tests in [#1422 (comment)](https://github.com/nginxinc/nginx-gateway-fabric/pull/1422#issuecomment-1888612433) make sense to you? cc @kate-osborn @pleshakov Thanks!

Sorry for the delay, @kevin85421. The test results look good! Just approved ✅ Do you mind rebasing?

kevin85421 commented:
> Do you mind rebasing?

Thank you for the prompt reply! I have already rebased the branch with the `main` branch.

kate-osborn commented on internal/mode/static/state/changed_predicate.go:
```diff
@@ -24,24 +24,12 @@ func (f funcPredicate) delete(object client.Object) bool {
-
-func (generationChangedPredicate) delete(_ client.Object) bool { return true }
+// FIXME(kevin85421): We should remove this predicate and update changeTrackingUpdater once #1432 is merged.
+type alwaysTruePredicate struct{}
```
Yeah, that's a better name than my suggestion.

kevin85421 replied:
Updated https://github.com/nginxinc/nginx-gateway-fabric/pull/1422/commits/1612d60bc3cb7a3be0f7aca37259b4a8201fc556
